node/src/crypto/hash.rs

```diff
@@ -172,6 +174,92 @@ impl From<Blake2bHash> for Digest {
     }
 }
 
+/// Sentinel hash to be used for hashing options in the case of [None].
+pub const SENTINEL0: Digest = Digest([0u8; 32]);
+/// Sentinel hash to be used by [hash_slice_rfold]. Terminates the fold.
+pub const SENTINEL1: Digest = Digest([1u8; 32]);
+/// Sentinel hash to be used by [hash_vec_merkle_tree] in the case of an empty list.
+pub const SENTINEL2: Digest = Digest([2u8; 32]);
+
+/// Hashes a pair of [`Digest`]s.
+pub fn hash_pair(hash1: &Digest, hash2: &Digest) -> Digest {
+    let mut to_hash = [0; Digest::LENGTH * 2];
+    to_hash[..Digest::LENGTH].copy_from_slice(&(hash1.to_array())[..]);
+    to_hash[Digest::LENGTH..].copy_from_slice(&(hash2.to_array())[..]);
+    hash(&to_hash)
+}
+
+/// Hashes a [`Vec`] of [`Digest`]s into a single [`Digest`] by constructing a [Merkle tree][1].
+/// Reduces pairs of elements in the [`Vec`] by repeatedly calling [hash_pair].
+/// This hash procedure is suited to hashing [BTree]s.
+///
+/// The pattern of hashing is as follows.  It is akin to [graph reduction][2]:
+///
+/// ```text
+/// a b c d e f
+/// |/  |/  |/
+/// g   h   i
+/// | /   /
+/// |/   /
+/// j   k
+/// | /
+/// |/
+/// l
+/// ```
+///
+/// Returns the empty [Digest] when the input is empty.
+///
+/// [1]: https://en.wikipedia.org/wiki/Merkle_tree
+/// [2]: https://en.wikipedia.org/wiki/Graph_reduction
+pub fn hash_vec_merkle_tree(vec: Vec<Digest>) -> Digest {
+    vec.into_iter()
+        .tree_fold1(|x, y| hash_pair(&x, &y))
+        .unwrap_or(SENTINEL2)
+}
+
+/// Hashes a [BTreeMap].
+pub fn hash_btree_map<K, V>(btree_map: &BTreeMap<K, V>) -> Result<Digest, bytesrepr::Error>
+where
+    K: ToBytes,
+    V: ToBytes,
+{
+    let mut kv_hashes: Vec<Digest> = Vec::with_capacity(btree_map.len());
+    for (key, value) in btree_map.iter() {
+        kv_hashes.push(hash_pair(&hash(key.to_bytes()?), &hash(value.to_bytes()?)))
+    }
+    Ok(hash_vec_merkle_tree(kv_hashes))
+}
+
+/// Hashes a `&[Digest]` using a [right fold][1].
+///
+/// This pattern of hashing is as follows:
+///
+/// ```text
+/// hash_pair(a, &hash_pair(b, &hash_pair(c, &SENTINEL)))
+/// ```
+///
+/// Unlike Merkle trees, this is suited to hashing heterogeneous lists we may wish to extend in the
+/// future (ie, hashes of data structures that may undergo revision).
+///
+/// Returns [`SENTINEL1`] when given an empty [`Vec`] as input.
+///
+/// [1]: https://en.wikipedia.org/wiki/Fold_(higher-order_function)#Linear_folds
+pub fn hash_slice_rfold(slice: &[Digest]) -> Digest {
+    slice
+        .iter()
+        .rfold(SENTINEL1, |prev, next| hash_pair(next, &prev))
```

Comment:
I just realized this could be simplified. (Not necessarily in this PR; maybe after fast sync has fully merged.)

```suggestion
    hash_slice_with_proof(slice, SENTINEL1)
```

Reply:
I have nothing against doing it right away :+1: